_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+[out](https://github.com/espressif/esp-idf/blob/master/components/esp_common/include/esp_err.h#L104-L124)
+[error](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_error.h#L191-L202)
+[handling](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/sys/__assert.h#L34-L89).
+
+In my opinion, disabling asserts for any reason other than to check how much
+codesize asserts take up in the system is more than likely a **reckless**
+endeavor, especially if they are used correctly.
+
+### Article Disclaimers
+
+- This post is about runtime asserting, not compile time asserting, such as
+  [`static_assert`](https://en.wikichip.org/wiki/c/assert.h/static_assert).
+- This post is targeting the ever-growing number of **non-safety critical**
+  embedded systems, such as consumer electronics, Class 1 medical devices, home
+  automation devices, etc.
+- The above types of embedded systems are primarily constrained by memory and
+  code space limitations, so the information below takes this into account.
+- The examples below are implemented for the ARM Cortex-M series of MCU's. Many
+  of the concepts can be generalized to other architectures.
+
+## Making the Most of Asserts
+
+To get the most value from asserting, it is important to set up some postmortem
+infrastructure. When an assert takes place (or even a system fault), there is _a
+lot_ of context and data one can capture to help debug these issues without a
+debugger attached to the device. These include but are not limited to:
+
+- the file and line number the assert took place,
+- the registers at the time of assert
+- values or function pointers on the stack, to help reassemble the backtrace
+- the logs that took place before the assert and possibly after the reboot took
+  place
+- for those using MyNewt[^1], ESP-IDF[^2], or have Memfault[^3] integrated, a
+  coredump can be captured at the time of the assert, which will provide a full
+  dump of system memory and registers.
+
+With all of the above, especially coredumps, it becomes advantageous to add as
+many asserts to the system as possible, especially for capturing those 1 in
+10,000 hour bugs. This allows developers to quickly track down and investigate
+bugs affecting users in the field, all without having to reproduce the issue on
+a development board.
+
+## Common Usages of Asserts
+
+Every firmware has places where it is and isn't appropriate to assert. The
+examples below likely apply to most non-safety-critical systems.
+
+### When malloc shouldn't fail
+
+Dynamic memory allocation is common in embedded systems, especially in today's
+Internet connected consumer electronics. These devices need to be able to
+receive an influx Bluetooth packets, use large HTTP buffers, and possibly load
+assets or fonts into RAM for a user interface. Keeping around statically
+allocated buffers for all of these functions would be irresponsible, especially
```

Comment:
s/irresponsible/wasteful 

irresponsible feels overly judgy and negative.

Reply:
Great point. Thanks for that edit.